app/workers/work-failer.ts

```diff
@@ -38,82 +38,121 @@ export default class WorkFailer implements Worker {
-            const key = `${jobId}${item.serviceID}`;
-            const message = failedMessage(item.id, jobServiceThresholds[key]);
-            workItemLogger.debug(message);
-            return handleWorkItemUpdate(
-              { workItemID: item.id, status: WorkItemStatus.FAILED,
-                scrollID: item.scrollID, hits: null, results: [], totalItemsSize: item.totalItemsSize,
-                errorMessage: message,
-              },
-              null,
-              workItemLogger);
-          }));
-        } catch (e) {
-          this.logger.error(`Error attempting to process work item updates for job ${jobId}.`);
-          this.logger.error(e);
+    }
+
+    return {
+      workItems: expiredWorkItems,
+      jobServiceThresholds: jobServiceThresholds,
+      maxId: maxId,
+    };
+  }
+
+  /**
+   * Find work items that're older than lastUpdateOlderThanMinutes and call handleWorkItemUpdate.
+   * @param lastUpdateOlderThanMinutes - upper limit on the duration since the last update
+   * @returns Resolves when the request is complete
+   */
+  async handleWorkItemUpdates(lastUpdateOlderThanMinutes: number): Promise<void> {
+    let done = false;
+    let startingId = 0;
+    const batchSize = env.workFailerBatchSize || 1000;
+    this.logger.info('Work failer processing started.');
+
+    while (!done) {
+      const { workItems, jobServiceThresholds, maxId: newId } = await this.getExpiredWorkItems(lastUpdateOlderThanMinutes,  startingId, batchSize);
+
+      if (newId > startingId) {
+        if (workItems.length > 0) {
+          for (const workItem of workItems) {
+            this.logger.warn(`expiring work item ${workItem.id}`, { workItemId: workItem.id });
+          }
+          const jobIds = new Set(workItems.map((item) => item.jobID));
+          for (const jobId of jobIds) {
+            jobIds.add(jobId);
+            try {
+              const itemsForJob = workItems.filter((item) => item.jobID === jobId);
+              await Promise.all(itemsForJob.map((item) => {
+                const workItemLogger = this.logger.child({ workItemId: item.id });
+                const key = `${jobId}${item.serviceID}`;
+                const message = failedMessage(item.id, jobServiceThresholds[key]);
+                workItemLogger.debug(message);
+                return handleWorkItemUpdateWithJobId(
+                  jobId,
+                  {
+                    workItemID: item.id, status: WorkItemStatus.FAILED,
+                    scrollID: item.scrollID, hits: null, results: [], totalItemsSize: item.totalItemsSize,
+                    errorMessage: message,
+                  },
+                  null,
+                  workItemLogger);
+              }));
+            } catch (e) {
+              this.logger.error(`Error attempting to process work item updates for job ${jobId}.`);
+              this.logger.error(e);
+            }
+          }
+
+          if (newId < (startingId + batchSize)) {
```

Comment:
I just wrote up a long message and realized I misinterpreted and this an optimization to let you make one fewer query if you are guaranteed there are no more items.

I was thinking a different way would be to check if the number of items returned was less than the batch size. I guess all the code is written to assume we're sorting on id so the current check will always be fine too.

Reply:
Correct.